nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java

```diff
@@ -0,0 +1,330 @@
+import org.apache.nifi.elasticsearch.ElasticSearchError;
+import org.apache.nifi.elasticsearch.IndexOperationRequest;
+import org.apache.nifi.expression.ExpressionLanguageScope;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.processors.elasticsearch.put.FlowFileJsonDescription;
+import org.apache.nifi.processors.elasticsearch.put.JsonProcessingError;
+import org.apache.nifi.util.StringUtils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@Tags({"json", "elasticsearch", "elasticsearch5", "elasticsearch6", "put", "index"})
+@CapabilityDescription("This processor puts user-supplied JSON into ElasticSearch. It does not require a schema.")
+public class PutElasticsearchJson extends AbstractProcessor implements ElasticSearchRestProcessor {
+
+    static final PropertyDescriptor ID_ATTRIBUTE = new PropertyDescriptor.Builder()
+        .name("put-es-json-id-attribute")
+        .displayName("ID Attribute")
+        .description("The attribute to use for setting the document ID in ElasticSearch. If the payload is an array, " +
+                "and this option is used for getting the ID, an exception will be raised.")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .addValidator(Validator.VALID)
+        .build();
+    static final PropertyDescriptor ID_JSON_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-json-id-json-path")
+        .displayName("ID JSONPath")
+        .description("If set, the document ID will be pulled from each JSON block using this JSONPath operation.")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+        .addValidator(Validator.VALID)
+        .build();
+    static final PropertyDescriptor INDEX_JSON_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-json-index-json-path")
+        .displayName("Index JSONPath")
+        .description("If set, the document's target index will be pulled from each JSON block using this JSONPath operation. " +
+                "If it is left blank or fails to evaluate to a valid value, the index property's valuable will be used instead. ")
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
```

Comment:
I might've missed this as part of a previous discussion, but is there a reason we wouldn't support FlowFile attributes for ID JSONPath, Index JSONPath, etc.? I see that they're being set in the OnScheduled, but do they need to be? Seems like even for arrays you could use a JSONPath defined for the flow file, for each element in the array?

Reply:
I was probably thinking about caching the compiled JsonPath objects. Probably not that big of a deal.

Reply:
When I push, that'll be refactored to be per-flowfile.